padlockcloud/sender.go

```diff
@@ -18,27 +18,48 @@ type EmailConfig struct {
 	Port string `yaml:"port"`
 	// Password used for authentication with the mail server
 	Password string `yaml:"password"`
+	// Sender mail address for outgoing mails. If empty, `User` is used instead.
+	From string `yaml:"from"`
 }
 
 // EmailSender implements the `Sender` interface for emails
 type EmailSender struct {
 	Config *EmailConfig
+	// Function used to actually send the mail. Same signature as `smtp.SendMail`.
+	SendFunc func(addr string, a smtp.Auth, from string, to []string, msg []byte) error
 }
 
-// Attempts to send an email to a given recipient. Through `smpt.SendMail`
+// NewEmailSender returns an EmailSender which sends mail using `smtp.SendMail`.
+// Its configuration points to the given `EmailConfig`.
+func NewEmailSender(c *EmailConfig) *EmailSender {
+	return &EmailSender{
+		Config:   c,
+		SendFunc: smtp.SendMail,
```

Comment:
Not sure if I'm a fan of this approach. I understand that you did this to make it more testable but I'm not sure if it it's not more trouble than it's worth. This is only to test the few lines in `EmailSender::Send`, correct?

Reply:
I could move the SendFunc in a global variable, like

`var smtpSendMail = smtp.SendMail`

and modify it in tests. I like it a bit less (because globals) but it keeps the code here slightly simpler. 

I think there can be some value in making this testable now while the implementation is simple, as it will make it more likely than future changes are tested. One thing I would like to add here and there, for instance, is better error reporting when the configuration is not valid.

What part is troublesome in your opinion? I could make SendFunc unexported if you don't like it being part of the 'public' part of EmailSender. 

(Anyway—you are the owner here. If you don't like this, I will revert it in the PR and probably keep it in my fork. I guess that's what forks are for :-))

Reply:
I guess what bothers me is that we're adding another layer of abstraction in addition to the `Sender` interface, but I realise that with the current implementation, this is the only way to really test the logic within `EmailSender`. This would be less of a big deal if it wouldn't require the addition of the `NewEmailSender` factory. At the very least, `SendFunc` should be a private member, but I think I would actually prefer stubbing out `smtp.SendMail` via a private global.

Reply:
Actually, now that I think about it, maybe this
```
	from := sender.Config.From
	if from == "" {
		from = sender.Config.User
	}
```
Shouldn't be in EmailSender::Send at all, but rather be handled during the initialisation process in within `cli.go` or `server.go`, in the form of:
```
if emailConfig.From == "" {
    emailConfig.From == emailConfig.User
}
```
Which would be much easier to test. The auth test are, as you said, too brittle to be of any real value anyway, so I'd just remove them.